src/rt/minfo.d

```diff
@@ -62,174 +62,382 @@ struct ModuleGroup
+                    }
+                }
 
-                        if (m.importedModules.length)
-                        {
-                            /* Internal runtime error, recursion exceeds number of modules.
-                             */
-                            (stackidx < _modules.length) || assert(0);
-
-                            // recurse
-                            stack[stackidx++] = StackRec(mods, idx);
-                            idx  = 0;
-                            mods = m.importedModules;
-                        }
+                // next dependency
+                ++sp.curDep;
+            }
+
+            // no cycles seen
+            return -1;
+        }
+
+        // The list of constructors that will be returned by the sorting.
+        immutable(ModuleInfo)*[] ctors;
+        // current element being inserted into ctors list.
+        size_t ctoridx = 0;
+
+
+        // This function will determine the order of construction/destruction and
+        // check for cycles. If a cycle is found, the cycle path is transformed
+        // into a string and thrown as an error.
+        //
+        // Each call into this function is given a module that has static
+        // ctor/dtors that must be dealt with. It recurses only when it finds
+        // dependencies that also have static ctor/dtors.
+        void processMod(int curidx)
+        {
+            assert(curidx != -1);
+            immutable ModuleInfo* current = _modules[curidx];
+
+            // First, determine what modules are reachable.
+            auto reachable = cast(size_t *)alloca(flagbytes);
+            auto cycleIdx = findDeps(curidx, reachable);
+            if(cycleIdx != -1)
+            {
+                auto cycleMod = _modules[cycleIdx];
+                // found a cycle
+                println("Cyclic dependency between module ", cycleMod.name, " and ", current.name);
+                genPath(cycleIdx, curidx);
+
+                foreach(midx; cyclePath[0 .. $-1])
+                {
+                    println(_modules[midx].name, bt(relevant, midx) ? "* ->" : " ->");
+                }
+                println(cycleMod.name, "*");
+                throw new Error(errmsg);
+            }
+
+            // process the dependencies. First, we process all relevant ones
+            bts(ctorstart, curidx);
+            foreach (int i; 0 .. len)
+            {
+                if(i != curidx && bt(reachable, i) &&
+                   bt(relevant, i) && !bt(ctordone, i))
+                {
+                    assert(!bt(ctorstart, i)); // sanity check, this should have been flagged a cycle earlier
+                    processMod(i);
+                }
+            }
+
+            // now mark this node, and all nodes reachable from this module as done.
+            bts(ctordone, curidx);
+            btr(ctorstart, curidx);
+            foreach (int i; 0 .. len)
+            {
+                if(bt(reachable, i))
```

Comment:
There isn't much to fix, b/c we have a graph with cycles between modules w/o ctor, a cheap O(N) sorting for DAGs doesn't work. We do a lot of caching so the normal case should be fairly linear, but the worst case has O(N) recursions, each of which reaching all the remaining modules through importing the next, so it's O(N^2) for the reachability computation.
Sorting Phobos took some 10s of microseconds last time I measured.